lib/timers.js

```diff
@@ -10,36 +10,84 @@ const kOnTimeout = Timer.kOnTimeout | 0;
 // Timeout values > TIMEOUT_MAX are set to 1.
 const TIMEOUT_MAX = 2147483647; // 2^31-1
 
-// IDLE TIMEOUTS
+
+// HOW and WHY the timers implementation works the way it does.
+//
+// Since many parts of Node.js and user applications rely on timeouts, there may
+// be a very large amount of timeouts scheduled at any given time.
+// As such, it is very important that the timers implementation be very fast and
+// efficent.
+//
+// Note: It is suggested you first read though the lib/internal/linkedlist.js
+// linked list implementaion. How it works is somewhat counter-intuitive to
+// most JavScript code, and timers depend on it extensively.
 //
-// Because often many sockets will have the same idle timeout we will not
-// use one timeout watcher per item. It is too much overhead.  Instead
-// we'll use a single watcher for all sockets with the same timeout value
-// and a linked list. This technique is described in the libev manual:
+// To achieve a high level of efficiency, the implementation is as distributed
+// and lazy as possible.
+//
+// Object maps are kept which contain linked lists keyed by their duration in
+// milliseconds.
+// The linked lists within are TimerWrap C++ handles with a linked list appended
+// to their JavaScript representation.
```

Comment:
> ```
// The map associates millisecond durations with linked lists. The linked lists
// are lead by a Timer instance, which is backed by a C++ TimerWrap, and
// subsequent items are instances of Timeout. An example in pseudocode:
//
//     map[1100] = Timer → Timeout → Timeout → Timeout
```

I don't think that's actually correct though.

```js
var list = {}
L.init(list)
// list { _idleNext: list, _idlePrev: list }

L.peek(list)
// > null
```

i.e. it's more like:

```js
// map[50] = ( TimerWrap ) .peek() / .shift() -> Timeout -> Timeout -> Timeout
```

The object you're getting the timers of off is always a TimerWrap at the start, but even though when "empty" the properties will point to itself, it will return null.

Not sure if that is important or how to articulate it better..